src/main/java/org/apache/ibatis/type/ArrayTypeHandler.java

```diff
@@ -30,9 +31,22 @@ public ArrayTypeHandler() {
     super();
   }
 
+  @Override
+  protected void setNullParameter(PreparedStatement ps, int i, JdbcType jdbcType) throws SQLException {
+    ps.setNull(i, Types.ARRAY);
+  }
+  
   @Override
   public void setNonNullParameter(PreparedStatement ps, int i, Object parameter, JdbcType jdbcType) throws SQLException {
-    ps.setArray(i, (Array) parameter);
+    Array array = null;
+    if (parameter instanceof Array) {
+      array = (Array)parameter;
+    }
+    else {
+      Object[] values = (Object[])parameter;
```

Comment:
Would check parameter is really instance of object array
and if not: throw TypeException("ArrayType Handler requires SQL array or java array parameter and does not support type " +parameter.getClass())
Otherwise a ClassCastException would be thrown

Beyond origin improvement, one might also consider as a next step: adding a case to handle java Collection <--> SQL array, or write a specific handler for that